src/cloudai/reports/training/parser.py

```
@@ -0,0 +1,215 @@
+        steps: list[TrainingStep] = self._build_steps(self._read_scalars(tr))
+        config: TrainingConfig = self._build_config(self.get_config(tr), tr, system)
+        return TrainingResults(config=config, steps=steps)
+
+    def _build_steps(self, scalars: list[Scalar]) -> list[TrainingStep]:
+        """list[Scalar] -> list[TrainingStep]: one TrainingStep per step (fields resolved via STEP_MAPPING)."""
+        steps: list[TrainingStep] = []
+        for step, step_scalars in self._group_by_step(scalars).items():
+            scalars_dict = {scalar.tag: scalar.value for scalar in step_scalars}
+            missing = [tag for tag in self.STEP_MAPPING.values() if tag not in scalars_dict]
+            if missing:
+                logging.warning(f"{type(self).__name__}: skipping step {step}, missing TB tags {missing}")
```

Ruff (0.15.18)

[warning] Logging statement uses f-string

(G004)